FF124 shadowRootClonable and friends (#33325)
mdn/content: files/en-us/web/html/element/template/index.md

      - : Hides the internal shadow root DOM from JavaScript.

    > **Note:** The value is reflected by {{domxref("HTMLTemplateElement.shadowRootMode")}}.
    > This is a feature of the HTML parser that cannot be used post-parsing by setting the `shadowrootmode` attribute through JavaScript.

>     > Only allowed values will create the shadow root; any other values, including empty ones, won't trigger this behavior.

Author: Is this true? I think it means you can't use `HTMLTemplateElement.shadowRootMode` to change the value of the root mode once it has been create - right?

But can you create an HTMLTemplateElement yourself, set the mode, and then add it to the DOM?

Author: Note, same issue as in https://github.com/mdn/content/pull/33325/files#r1590568866

    > **Note:** You may find the non-standard `shadowroot` attribute in older tutorials and examples that used to be supported in Chrome 90-110. This attribute has since been removed and replaced by the standard `shadowrootmode` attribute.

- `shadowrootdelegatesfocus`

  - : Sets the value of the [`delegatesFocus`](/en-US/docs/Web/API/ShadowRoot/delegatesFocus) property of a [`ShadowRoot`](/en-US/docs/Web/API/ShadowRoot) created using this element to `true`.

>     If this is set and a non-focussable element in the shadow tree is selected, then focus is delegated to the first focussable element in the tree.

Reviewer: ```suggestion
    If this is set and a non-focusable element in the shadow tree is selected, then focus is delegated to the first focusable element in the tree.
```

    If set, a clone of the shadow host (the parent element of this `<template>`) created with {{domxref("Node.cloneNode()")}} or {{domxref("Document.importNode()")}} will include a shadow root in the copy.
    The value defaults to `false` and is reflected by {{domxref("HTMLTemplateElement.shadowRootClonable")}}.

## Usage notes

There are two main ways to use the `<template>` element.

### Template document fragment

By default, the element's content is not rendered, but is parsed into a [document fragment](/en-US/docs/Web/API/DocumentFragment).
Using the {{domxref("HTMLTemplateElement.content", "content")}} property in JavaScript, this fragment can be cloned via the {{domxref("Node.cloneNode", "cloneNode")}} method and inserted into the DOM.

### Declarative Shadow DOM

If the element contains the [`shadowrootmode`](#shadowrootmode) attribute, the HTML parser will immediately generate a shadow DOM. The element is replaced in the DOM by its content wrapped in a [shadow root](/en-US/docs/Glossary/Shadow_tree), which is attached to the parent element.

> This is the declarative equivalent to calling {{domxref("Element.attachShadow()")}} to attach a shadow root to an element.

Reviewer: ```suggestion
This is the declarative equivalent of calling {{domxref("Element.attachShadow()")}} to attach a shadow root to an element.
```

    > **Note:** You may find the non-standard `shadowroot` attribute in older tutorials and examples that used to be supported in Chrome 90-110. This attribute has since been removed and replaced by the standard `shadowrootmode` attribute.

- `shadowrootdelegatesfocus`

>   - : Xxxxx

Author: TBD

- `shadowrootclonable`

  - : Sets the value of [`clonable`](/en-US/docs/Web/API/ShadowRoot/clonable) to `true` on a shadow root that has been created declaratively (by specifying the [`shadowrootmode`](#shadowrootmode) attribute).
    If it is declared, a clone of the shadow host (parent element of this `<template>`) created with {{domxref("Node.cloneNode()")}} or {{domxref("Document.importNode()")}} will include a shadow root in the copy.
    The value defaults to `false` and is reflected by {{domxref("HTMLTemplateElement.shadowRootClonable")}}.

## Usage notes

There are two main ways to use the `<template>` element.

### Template document fragment

By default, the element's content is not rendered, but is parsed into a [document fragment](/en-US/docs/Web/API/DocumentFragment).
Using the {{domxref("HTMLTemplateElement.content", "content")}} property in JavaScript, this fragment can be cloned via the {{domxref("Node.cloneNode", "cloneNode")}} method and inserted into the DOM.

> ### Declarative Shadow DOM

Author: FYI I pulled out this topic, but  it isn't visible enough. I think we need a topic just on this at guide level. OK for now maybe

Author: Ah, found  https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_shadow_DOM#declaratively_with_html in the guide. I'll cross link.

{{EmbedLiveSample('Avoiding_DocumentFragment_pitfall')}}

> ### Declarative Shadow DOM with delegated focus

Author: FYI this started out as a live example but unfortunately this does not work in MDN environment with `<template>`. Its possible to do this programmatically, but I am unlikely to have time this time round.

Reviewer: The demo seems fine to me; it makes sense.

One thing I did notice — you've put this section as a child of the "Avoiding DocumentFragment pitfalls" H2, rather than the "Examples" H2. I'm pretty sure that's not where you meant to put it?

{{EmbedLiveSample('Avoiding_DocumentFragment_pitfall')}}

### Declarative Shadow DOM with delegated focus

This example demonstrates how `shadowrootdelegatesfocus` is applied to a shadow root that is created declaratively, and the effect this has on focus.

> The code first declares a shadow roots inside a `<div>` elements, using the `<template>` element with the `shadowrootmode` attribute.

Reviewer: ```suggestion
The code first declares a shadow root inside a `<div>` element, using the `<template>` element with the `shadowrootmode` attribute.
```

> This displays both a non-focussable `<div>` containing text and a focussable `<input>` element.

Reviewer: ```suggestion
This displays both a non-focusable `<div>` containing text and a focusable `<input>` element.
```

It also uses CSS to style elements with [`:focus`](/en-US/docs/Web/CSS/:focus) to blue, and to set the normal styling of the host element.

```html
<div>
  <template shadowrootmode="open">
    <style>
      :host {
        display: block;
        border: 1px dotted black;
        padding: 10px;
        margin: 10px;
      }
      :focus {
        outline: 2px solid blue;
      }
    </style>
    <div>Clickable Shadow DOM text</div>
    <input type="text" placeholder="Input inside Shadow DOM" />
  </template>
</div>
```

> The second code block is identical except that it sets the `shadowrootdelegatesfocus`, which delegates focus to the first focussable element in the tree if a non-focussable element in the tree is selected.

Reviewer: ```suggestion
The second code block is identical except that it sets the `shadowrootdelegatesfocus` attribute, which delegates focus to the first focusable element in the tree if a non-focusable element in the tree is selected.
```

```html
<div>
  <template shadowrootmode="open" shadowrootdelegatesfocus>
    <style>
      :host {
        display: block;
        border: 1px dotted black;
        padding: 10px;
        margin: 10px;
      }
      :focus {
        outline: 2px solid blue;
      }
    </style>
    <div>Clickable Shadow DOM text</div>
    <input type="text" placeholder="Input inside Shadow DOM" />
  </template>
</div>
```

Last of all we use the following CSS to apply a green-yellow border to the parent `<div>` element when it has focus.

```css
div:focus {

>   border: 1px solid greenyellow;

Reviewer: I would choose a darker color for this. For a little while, I thought there was something wrong with the demo or the screenshot, as I couldn't see the div border.

Author: Done - changed to red.

}
```

The results are shown below.
When the HTML is first rendered, the elements have no styling, as shown in the first image.
For the shadow root that does not have `shadowrootdelegatesfocus` set you can click anywhere except the `<input>` and the focus does not change (if you select the `<input>` element it will look like the second image).

![Screenshot of code with no focus set](template_with_no_focus.png)

> For the shadow root with `shadowrootdelegatesfocus` set, clicking on the text (which is non-focussable) selects the input element, as this is the first focussable element in the tree.

Reviewer: ```suggestion
For the shadow root with `shadowrootdelegatesfocus` set, clicking on the text (which is non-focusable) selects the `<input>` element, as this is the first focusable element in the tree.
```
